File: scripts/build_global_review_queue.py
#!/usr/bin/env python3
"""
Build machine-readable and human-readable review queues for cleaned global
country datasets.
"""

from __future__ import annotations

import csv
import json
import re
import sys
from collections import Counter
from datetime import UTC, datetime
from pathlib import Path


REPO_ROOT = Path("/Users/joseph/GIT/places-of-worship")

PLACEHOLDER_PATTERN = re.compile(r"^Place of Worship \d+$")
GENERIC_LABEL_PATTERN = re.compile(
    r"^(Christian|Anglican|Roman_Catholic|Jewish|Sikh|Mormon|Methodist|Lutheran) Place of Worship$",
    re.IGNORECASE,
)
INSTITUTIONAL_PATTERN = re.compile(r"\b(academy|school|seminary|college)\b", re.IGNORECASE)
RETREAT_PATTERN = re.compile(r"\b(retreat|prayer|meditation)\b", re.IGNORECASE)
HALL_CENTRE_HOUSE_PATTERN = re.compile(r"\b(centre|center|hall|house|community)\b", re.IGNORECASE)

CATEGORY_ORDER = {
    "placeholder_name": 1,
    "generic_worship_label": 1,
    "institutional_site": 1,
    "retreat_or_prayer_site": 2,
    "hall_centre_house_site": 2,
    "missing_core_tags": 3,
}


def latest_snapshot_dir() -> Path:
    base_dir = REPO_ROOT / "data" / "intermediate" / "global"
    snapshot_dirs = [path for path in base_dir.iterdir() if path.is_dir()]
    if not snapshot_dirs:
        raise FileNotFoundError(f"No snapshot directories found under {base_dir}")
    return sorted(snapshot_dirs)[-1]


def iter_cleaned_files(input_path: Path) -> list[Path]:
    if input_path.is_file():
        return [input_path]
    return sorted(input_path.glob("*_places_cleaned.json"))


def is_weak_tag_record(record: dict) -> bool:
    tags = record.get("tags_raw") or {}
    return tags.get("amenity") in (None, "place_of_worship") and tags.get("building") in (None, "yes")


def classify_record(record: dict) -> tuple[str, int] | None:
    tags = record.get("tags_raw") or {}
    amenity = tags.get("amenity")
    building = tags.get("building")
    name = record.get("name") or ""

    weak = is_weak_tag_record(record)

    if weak and PLACEHOLDER_PATTERN.search(name):
        return "placeholder_name", CATEGORY_ORDER["placeholder_name"]

    if weak and GENERIC_LABEL_PATTERN.search(name):
        return "generic_worship_label", CATEGORY_ORDER["generic_worship_label"]

    if INSTITUTIONAL_PATTERN.search(name):
        return "institutional_site", CATEGORY_ORDER["institutional_site"]

    if RETREAT_PATTERN.search(name):
        return "retreat_or_prayer_site", CATEGORY_ORDER["retreat_or_prayer_site"]

    if weak and HALL_CENTRE_HOUSE_PATTERN.search(name):
        return "hall_centre_house_site", CATEGORY_ORDER["hall_centre_house_site"]

    if amenity is None and building is None:
        return "missing_core_tags", CATEGORY_ORDER["missing_core_tags"]

    return None


def build_queue(records: list[dict], country_code: str) -> list[dict]:
    queue = []

    for record in records:
        classification = classify_record(record)
        if classification is None:
            continue

        category, priority = classification
        tags = record.get("tags_raw") or {}

        queue.append(
            {
                "priority": priority,
                "category": category,
                "country_code": country_code,
                "id": record.get("id", ""),
                "name": record.get("name", ""),
                "religion": record.get("religion", ""),
                "denomination": record.get("denomination", ""),
                "address": record.get("address", ""),
                "amenity": tags.get("amenity", ""),
                "building": tags.get("building", ""),
                "lat": record.get("lat", ""),
                "lng": record.get("lng", ""),
                "tags_raw": json.dumps(tags, ensure_ascii=False, sort_keys=True),
            }
        )

    queue.sort(key=lambda row: (row["priority"], row["category"], row["name"], row["id"]))
    return queue


def write_country_queue(cleaned_file: Path, output_dir: Path) -> dict:
    country_code = cleaned_file.name.replace("_places_cleaned.json", "").upper()
    records = json.loads(cleaned_file.read_text(encoding="utf-8"))
    queue = build_queue(records, country_code)

    csv_path = output_dir / f"{country_code.lower()}_review_queue.csv"
    md_path = output_dir / f"{country_code.lower()}_review_queue.md"

    fieldnames = [
        "priority",
        "category",
        "country_code",
        "id",
        "name",
        "religion",
        "denomination",
        "address",
        "amenity",
        "building",
        "lat",
        "lng",
        "tags_raw",
    ]

    with csv_path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(queue)

    category_counts = Counter(row["category"] for row in queue)
    priority_counts = Counter(row["priority"] for row in queue)
    lines = [
        f"# {country_code} Review Queue",
        "",
        f"- cleaned dataset size: {len(records)}",
        f"- queued for manual review: {len(queue)}",
        f"- priority 1 records: {priority_counts.get(1, 0)}",
        f"- priority 2 records: {priority_counts.get(2, 0)}",
        f"- priority 3 records: {priority_counts.get(3, 0)}",
        "",
        "## Categories",
        "",
    ]

    for category in sorted(category_counts, key=lambda key: (CATEGORY_ORDER[key], key)):
        lines.append(f"- `{category}`: {category_counts[category]}")

    md_path.write_text("\n".join(lines) + "\n", encoding="utf-8")

    return {
        "country_code": country_code,
        "cleaned_file": str(cleaned_file.relative_to(REPO_ROOT)),
        "csv_path": str(csv_path.relative_to(REPO_ROOT)),
        "md_path": str(md_path.relative_to(REPO_ROOT)),
        "cleaned_count": len(records),
        "review_count": len(queue),
    }


def main() -> None:
    args = sys.argv[1:]
    positional = [arg for arg in args if not arg.startswith("--")]
    input_path = Path(positional[0]) if positional else latest_snapshot_dir()
    if not input_path.is_absolute():
        input_path = (REPO_ROOT / input_path).resolve()

    cleaned_files = iter_cleaned_files(input_path)
    if not cleaned_files:
        raise FileNotFoundError(f"No cleaned files found under {input_path}")

    snapshot_name = input_path.name if input_path.is_dir() else input_path.parent.name
    output_dir = REPO_ROOT / "docs" / "review_queues" / snapshot_name
    output_dir.mkdir(parents=True, exist_ok=True)

    results = [write_country_queue(cleaned_file, output_dir) for cleaned_file in cleaned_files]

    manifest = {
        "generated_at": datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "script": "scripts/build_global_review_queue.py",
        "input_path": str(input_path.relative_to(REPO_ROOT)),
        "countries": results,
    }

    manifest_path = output_dir / "review_queue_manifest.json"
    manifest_path.write_text(json.dumps(manifest, indent=2), encoding="utf-8")
    print(f"Wrote review queue manifest: {manifest_path}")


if __name__ == "__main__":
    main()

File: scripts/build_global_review_queue.R
# Global review queue builder
# Usage:
#   Rscript scripts/build_global_review_queue.R [input_path]
#
# Example:
#   Rscript scripts/build_global_review_queue.R data/intermediate/global/undated

suppressPackageStartupMessages({
  library(jsonlite)
  library(purrr)
})

category_order <- c(
  "placeholder_name" = 1,
  "generic_worship_label" = 1,
  "institutional_site" = 1,
  "retreat_or_prayer_site" = 2,
  "hall_centre_house_site" = 2,
  "missing_core_tags" = 3
)

parse_args <- function(args) {
  positional <- args[!startsWith(args, "--")]

  list(
    input_path = if (length(positional) >= 1) positional[[1]] else NULL
  )
}

find_latest_snapshot_dir <- function() {
  base_dir <- file.path("data", "intermediate", "global")
  if (!dir.exists(base_dir)) {
    stop("No intermediate global directory found at ", base_dir, ".")
  }

  snapshot_dirs <- list.dirs(base_dir, recursive = FALSE, full.names = TRUE)
  if (length(snapshot_dirs) == 0) {
    stop("No snapshot directories found under ", base_dir, ".")
  }

  snapshot_dirs[[which.max(basename(snapshot_dirs))]]
}

iter_input_files <- function(input_path) {
  if (file.exists(input_path) && !dir.exists(input_path)) {
    return(input_path)
  }

  deduplicated_files <- sort(list.files(
    input_path,
    pattern = "_places_deduplicated\\.json$",
    full.names = TRUE
  ))

  if (length(deduplicated_files) > 0) {
    return(deduplicated_files)
  }

  sort(list.files(
    input_path,
    pattern = "_places_cleaned\\.json$",
    full.names = TRUE
  ))
}

matches_pattern <- function(pattern, value) {
  grepl(pattern, value %||% "", perl = TRUE, ignore.case = TRUE)
}

is_weak_tag_record <- function(record) {
  tags <- record$tags_raw %||% list()
  amenity <- tags[["amenity"]] %||% NULL
  building <- tags[["building"]] %||% NULL

  (is.null(amenity) || identical(amenity, "place_of_worship")) &&
    (is.null(building) || identical(building, "yes"))
}

classify_record <- function(record) {
  tags <- record$tags_raw %||% list()
  amenity <- tags[["amenity"]] %||% NULL
  building <- tags[["building"]] %||% NULL
  name <- record$name %||% ""
  weak <- is_weak_tag_record(record)

  if (weak && matches_pattern("^Place of Worship \\d+$", name)) {
    return(list(category = "placeholder_name", priority = category_order[["placeholder_name"]]))
  }

  if (
    weak &&
      matches_pattern("^(Christian|Anglican|Roman_Catholic|Jewish|Sikh|Mormon|Methodist|Lutheran) Place of Worship$", name)
  ) {
    return(list(category = "generic_worship_label", priority = category_order[["generic_worship_label"]]))
  }

  if (matches_pattern("\\b(academy|school|seminary|college)\\b", name)) {
    return(list(category = "institutional_site", priority = category_order[["institutional_site"]]))
  }

  if (matches_pattern("\\b(retreat|prayer|meditation)\\b", name)) {
    return(list(category = "retreat_or_prayer_site", priority = category_order[["retreat_or_prayer_site"]]))
  }

  if (weak && matches_pattern("\\b(centre|center|hall|house|community)\\b", name)) {
    return(list(category = "hall_centre_house_site", priority = category_order[["hall_centre_house_site"]]))
  }

  if (is.null(amenity) && is.null(building)) {
    return(list(category = "missing_core_tags", priority = category_order[["missing_core_tags"]]))
  }

  NULL
}

queue_row_from_record <- function(record, country_code) {
  classification <- classify_record(record)
  if (is.null(classification)) {
    return(NULL)
  }

  tags <- record$tags_raw %||% list()
  ordered_tag_names <- sort(names(tags))
  ordered_tags <- tags[ordered_tag_names]

  list(
    priority = classification$priority,
    category = classification$category,
    country_code = country_code,
    id = record$id %||% "",
    name = record$name %||% "",
    religion = record$religion %||% "",
    denomination = record$denomination %||% "",
    address = record$address %||% "",
    amenity = tags[["amenity"]] %||% "",
    building = tags[["building"]] %||% "",
    lat = as.character(record$lat %||% ""),
    lng = as.character(record$lng %||% ""),
    tags_raw = as.character(toJSON(ordered_tags, auto_unbox = TRUE, null = "null"))
  )
}

build_queue <- function(records, country_code) {
  queue <- records |>
    map(\(record) queue_row_from_record(record, country_code)) |>
    compact()

  if (length(queue) == 0) {
    return(data.frame())
  }

  queue_df <- queue |>
    map(\(row) as.data.frame(row, stringsAsFactors = FALSE)) |>
    list_rbind()

  queue_df[order(queue_df$priority, queue_df$category, queue_df$name, queue_df$id), , drop = FALSE]
}

write_country_queue <- function(path, output_dir) {
  country_code <- toupper(
    sub(
      "_places_(deduplicated|cleaned)\\.json$",
      "",
      basename(path)
    )
  )

  records <- read_json(path, simplifyVector = FALSE)
  queue <- build_queue(records, country_code)

  csv_path <- file.path(output_dir, paste0(tolower(country_code), "_review_queue.csv"))
  md_path <- file.path(output_dir, paste0(tolower(country_code), "_review_queue.md"))

  write.csv(queue, csv_path, row.names = FALSE, na = "")

  category_counts <- if (nrow(queue) == 0) integer() else table(queue$category)
  priority_counts <- if (nrow(queue) == 0) integer() else table(queue$priority)

  lines <- c(
    paste("#", country_code, "Review Queue"),
    "",
    paste("- cleaned dataset size:", length(records)),
    paste("- queued for manual review:", nrow(queue)),
    paste("- priority 1 records:", priority_counts[["1"]] %||% 0),
    paste("- priority 2 records:", priority_counts[["2"]] %||% 0),
    paste("- priority 3 records:", priority_counts[["3"]] %||% 0),
    "",
    "## Categories",
    ""
  )

  if (length(category_counts) > 0) {
    ordered_categories <- names(category_counts)[order(
      unname(category_order[names(category_counts)]),
      names(category_counts)
    )]

    lines <- c(
      lines,
      map_chr(
        ordered_categories,
        \(category) sprintf("- `%s`: %s", category, category_counts[[category]])
      )
    )
  }

  writeLines(lines, md_path, useBytes = TRUE)

  list(
    country_code = country_code,
    input_file = path,
    csv_path = csv_path,
    md_path = md_path,
    input_count = length(records),
    review_count = nrow(queue)
  )
}

write_review_manifest <- function(results, output_dir, input_path) {
  manifest <- list(
    generated_at = format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC"),
    script = "scripts/build_global_review_queue.R",
    input_path = input_path,
    countries = results
  )

  manifest_path <- file.path(output_dir, "review_queue_manifest.json")
  write_json(manifest, manifest_path, pretty = TRUE, auto_unbox = TRUE, null = "null")
  message(sprintf("Wrote review queue manifest: %s", manifest_path))
}

main <- function() {
  args <- parse_args(commandArgs(trailingOnly = TRUE))
  input_path <- args$input_path %||% find_latest_snapshot_dir()

  input_files <- iter_input_files(input_path)
  if (length(input_files) == 0) {
    stop("No deduplicated or cleaned files found under ", input_path, ".")
  }

  snapshot_name <- if (dir.exists(input_path)) basename(input_path) else basename(dirname(input_path))
  output_dir <- file.path("docs", "review_queues", snapshot_name)
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

  results <- input_files |>
    map(\(path) write_country_queue(path, output_dir))

  write_review_manifest(results, output_dir, input_path)
}

main()
